vision/dcgan_mnist/dcgan.jl

```diff
@@ -0,0 +1,130 @@
+    image_array = @. Gray(image_array + 1f0) / 2f0
+    return image_array
+end
+
+# Loss functions
+function discriminator_loss(real_output, fake_output)
+    real_loss = mean(logitbinarycrossentropy.(real_output, 1f0))
+    fake_loss = mean(logitbinarycrossentropy.(fake_output, 0f0))
+    return real_loss + fake_loss
+end
+
+generator_loss(fake_output) = mean(logitbinarycrossentropy.(fake_output, 1f0))
+
+function train_discriminator!(gen, dscr, batch, opt_dscr, hparams)
+    noise = randn(Float32, hparams.latent_dim, hparams.batch_size) |> gpu
+    fake_input = gen(noise)
+    ps = Flux.params(dscr)
+    # Taking gradient
+    loss, back = Flux.pullback(ps) do
+        discriminator_loss(dscr(batch), dscr(fake_input))
+    end
+    grad = back(1f0)
+    update!(opt_dscr, ps, grad)
+    return loss
+end
+
+function train_generator!(gen, dscr, batch, opt_gen, hparams)
+    noise = randn(Float32, hparams.latent_dim, hparams.batch_size) |> gpu
+    ps = Flux.params(gen)
+    # Taking gradient
+    loss, back = Flux.pullback(ps) do
+        generator_loss(dscr(gen(noise)))
+    end
+    grad = back(1f0)
+    update!(opt_gen, ps, grad)
+    return loss
+end
+
+function train(; kws...)
+    # Model Parameters
+    hparams = HyperParams(; kws...)
+
+    # Load MNIST dataset
+    images, _ = MLDatasets.MNIST.traindata(Float32)
+    # Normalize to [-1, 1] and convert it to WHCN
+    image_tensor = permutedims(reshape(@.(2f0 * images - 1f0), 28, 28, 1, :), (2, 1, 3, 4))
+    # Partition into batches
+    data = [image_tensor[:, :, :, r] |> gpu for r in partition(1:60000, hparams.batch_size)]
+
+    fixed_noise = [randn(hparams.latent_dim, 1) |> gpu for _=1:hparams.output_x*hparams.output_y]
+
+    # Discriminator
+    dscr =  Chain(
+        Conv((4, 4), 1 => 64; stride = 2, pad = 1),
+        x->leakyrelu.(x, 0.2f0),
+        Dropout(0.25),
+        Conv((4, 4), 64 => 128; stride = 2, pad = 1),
+        x->leakyrelu.(x, 0.2f0),
+        Dropout(0.25), 
+        x->reshape(x, 7 * 7 * 128, :),
+        Dense(7 * 7 * 128, 1)) |> gpu
+
+    # Generator
+    gen = Chain(
+        Dense(hparams.latent_dim, 7 * 7 * 256),
+        BatchNorm(7 * 7 * 256, relu),
+        x->reshape(x, 7, 7, 256, :),
+        ConvTranspose((5, 5), 256 => 128; stride = 1, pad = 2),
+        BatchNorm(128, relu),
+        ConvTranspose((4, 4), 128 => 64; stride = 2, pad = 1),
+        BatchNorm(64, relu),
+        ConvTranspose((4, 4), 64 => 1, tanh; stride = 2, pad = 1),
+        ) |> gpu
+
+    # Optimizers
+    opt_dscr = ADAM(hparams.lr_dscr)
+    opt_gen = ADAM(hparams.lr_gen)
+
+    # Training
+    train_steps = 0
+    for ep in 1:hparams.epochs
+        @info "Epoch $ep"
+        for batch in data
+            # Update discriminator and generator
+            loss_dscr = train_discriminator!(gen, dscr, batch, opt_dscr, hparams)
+            loss_gen = train_generator!(gen, dscr, batch, opt_gen, hparams)
```

Comment:
here part of the computation done in train_discriminator! could be reused in train_generator!, i.e. in principle one need only a single forward of the generator.  

This is also done here https://pytorch.org/tutorials/beginner/dcgan_faces_tutorial.html

I don't know how to do it properly with zygote, needs some thinking. If you have a simple solution in mind  you could add this optimization, otherwise we can revisit in the future

Reply:
Unfortunately I don't know how to do it, too.